README.md
![SOF](sof_files/logo_blue.png)

# [Setor de Orçamento e Finanças do HUSM](http://www.sofhusm.net.br)

##### Requisitos

* Apache 2.4.18+
* PHP 7.0+
* MySQL 5.7+

##### PDF

Para geração de PDF, precisa criar um diretório na raiz do projeto com o nome 'temp'. É onde ficarão os arquivos temporários necessários para a criação e visualização dos PDFs gerados no site.



CHANGELOG.md
## CHANGELOG ##

#### v2.4.7 - 19/08/2019 ####
- Adiciona campo ano para parte de faturamento (aihs)
- Relatórios de AIHs mais filtratos
- Mostra ajustes SIA e SIH no relatório de faturamento
- Mostra sub totais no relatório de faturamento
- Carrega dinamicamento select com opções de contrato para fazer relatório (bug)

#### v2.4.6 - 02/08/2019 ####
- Formulário para faturamento aprovado
- Tabela para visualizar, editar e remover faturamentos aprovados
- Melhoria no relatório de receitas
- Visualização, edição e remoção das contratualizações

#### v2.4.5 - 24/07/2019 ####
- Organização da interface para AIHS
- Relatório de Receitas Recebidas com sub totais
- Correção de bug na inserção de observações (<= 300 caracteres)
- Cadastro de tipos para receitas

#### v2.4.4 - 07/07/2019 ####
- Implementação de tabela para visualização, edição e remoção de Receitas Recebidas (AIHS)

#### v2.4.3 - 28/06/2019 ####
- Implementação de formulário Receitas Recebidas (AIHS)
- Correção do relatório SIAFI (com opção: ignorar fontes)

#### v2.4.2 - 02/06/2019 ####
- Inicio de implementação de AIHs (somente formulário inicial)

#### v2.4.1 - 24/02/2019 ####
- Suporte para mais de uma nota por mês
- Organização da estrutura interna do sistema
- Relatório de contrato mostrando todas as notas
- Suporte para exclusão de contratos

#### v2.4.0 - 06/01/2019 ####
- Editar NFs lançadas
- Corrige bug o checkbox 'Reajuste' que não estava operando
- Adiciona grupo para o qual a mensalidade é resignada
- Adiciona filtros para geração de relatórios do contrato (com nota, paga, etc)
- Criação do banco de dados para 2019

#### v2.3.10 - 11/11/2018 ####
- Separa pedidos em vencimento por setor
- Adiciona novos campos (paga e aguarda orçamento) no cadastro de mensalidade de contratos

#### v2.3.9 - 01/09/2018 ####
- Adiciona vigência completa (início e fim) de contratos
- Relatório de contratos
- Ajustes na página inicial de contratos

#### v2.3.8 - 05/07/2018 ####
- Cadastra reajuste de mensalidade (opcional)
- Cadastra CNPJ da empresa

#### v2.3.7 - 17/06/2018 ####
- Adiciona nome do fornecedor nos pedidos em vencimento
- Funcionalidade para gerenciar contratos de empresas

#### v2.3.6 - 22/04/2018 ####
- Imprime lista de pedidos em vencimento com somatório e descrição do relatório
- Mostra somente pedidos em vencimento que estão em análise
- Elimina restrição de usuário na edição de pedidos reprovados e rascunhos
- Modifica importação de itens: apenas faz o upload do arquivo com os dados, fica mais rápido assim

#### v2.3.5 - 01/04/2018 ####
- Mostra os pedidos que irão vencer no mês atual ao invés dos pregões

#### v2.3.4 - 18/03/2018 ####
- Relatório de Processos não devolvidos (recepção)
- Alerta para vencimento de pregões no mês atual

#### v2.3.3 - 11/03/2018 ####
- Corrige bug ao editar empenhos
- Permite empresas acessarem o site para visualizarem arquivos

#### v2.3.2 - 13/02/2018 ####
- Corrige bug ao salvar informações no site
- Permite a alteração de pedidos mesmo após a análise (precisa alterar o status e reprovar o pedido). As informações salvas de empenho e fontes são mantidas. Ao reprovar o pedido basta voltar para o setor que fez o pedido e continuar a edição dos "Rascunhos". Quando o pedido for enviado ao SOF novamente ele estará "Em Análise".
- Adiciona nova página "Ações Sociais"
- Adiciona nova página "Depoimentos"
- Corrige bug ao imprimir determinados pedidos (caracteres inválidos)

#### v2.3.1 - 29/01/2018 ####
- Desvinculação de pedidos com fontes e transferências de saldos com fontes
- Melhoria do relatório de liberações orçamentárias (melhor descrição do relatório e retirada de campos redundantes)
- Realização de pedidos marcados com Plano de trabalho
- Ajustes no layout do relatório de pedidos
- Correção de bug no relatório de pedidos (mostrava pedidos que não eram do Setor selecionado)
- Possibilidade de desfazer liberação orçamentária (disponível somente ao usar o banco principal)

Nota: O banco principal (main) é o banco do ano corrente

#### v2.3.0 - 15/01/2018 ####
- Corrige bug ao editar itens (valor utilizado sempre incorreto)
- Opção "Todas" no campo de prioridades para relatório de pedidos
- Permite edição de itens cancelados e cancelar/descancear itens pela ferramenta de edição
- Corrige bug ao remover pedido sem itens
- Retira limite de linhas de relatório de pedidos e outros
- Finaliza relatório SIAFI com seleção de múltiplas fontes

#### v2.2.10 - 17/12/2017 ####
- Corrige bug de codificação de caracteres ao editar itens
- Corrige bug na remoção dos pedidos
- Cria mecanismo para alternar entre diferentes bancos dentro do site (2017 e 2018)
- Corrige relatório de usuários: mostra apenas os ativos no sistema.

#### v2.2.9 - 27/11/2017 ####
- Permite seleção de mais de um processo ao gerar relatório SIAFI.

#### v2.2.8 - 12/11/2017 ####
- Corrige inconsistência na informação de SIAFI no banco
- Muda formato de datas em itens para facilitar relatórios futuros
- Implementa relatório SIAFI por setor, fonte de recurso, número de processo e vigência

#### v2.2.7 - 04/11/2017 ####
- Adiciona prioridade "Hoje" para realização de pedidos.
- Formatação de relatório de pedidos (retirada de informações desnecessárias)

#### v2.2.6 - 29/10/2017 ####
- Corrige bug que zerava o saldo do sof e atualizava constantemente o saldo dos setores.
- Corrige bug ao reprovar pedido.
- Corrige bug ao mostrar o painel das solicitações (mensagem do DataTables)
- Coloca valor mínimo nas quantidades (de contrato e utilizada) ao editar item (evita conflito de informações)]
- Agrega relatório de fontes com relatório de pedidos.

#### v2.2.5 - 15/10/2017 ####
- Correção de bug ao editar itens
- Melhoria geral na edição e cadastro de itens

#### v2.2.4 - 11/09/2017 ####
- Remove possibilidade de desfazer liberação orçamentária.
- Insere novos objetos para representar liberações orçamentárias e pedido de contrato.
- Coleta de bugs e análise de performance (importação de itens ainda está instável)


import.js
const id_item_processo = 0;
const id_item_contrato = 1;
const cod_despesa = 2;
const descr_despesa = 3;
const descr_tipo_doc = 4;
const num_contrato = 5;
const num_processo = 6;
const descr_mod_compra = 7;
const num_licitacao = 8;
const dt_inicio = 9;
const dt_fim = 10;
const dt_geracao = 11;
const cgc_fornecedor = 12;
const nome_fornecedor = 13;
const num_extrato = 14;
const cod_estruturado = 15;
const nome_unidade = 16;
const cod_reduzido = 17;
const complemento_item = 18;
const descricao = 19;
const id_extrato_contr = 20;
const vl_unitario = 21;
const qt_contrato = 22;
const vl_contrato = 23;
const qt_utilizado = 24;
const vl_utilizado = 25;
const qt_saldo = 26;
const vl_saldo = 27;
const id_unidade = 28;
const ano_orcamento = 29;
const seq_item_processo = 32;

const CSVsaida = document.getElementById('CSVsaida');

const leitorDeCSV = new FileReader();

window.onload = function init() {
    leitorDeCSV.onload = leCSV;
}

function pegaCSV(inputFile) {
    CSVsaida.innerHTML += "<p>Novo arquivo selecionado</p>";
    const file = inputFile.files[0];
    leitorDeCSV.readAsText(file);
}

function mysql_real_escape_string(str) {
    return str.replace(/[\0\x08\x09\x1a\n\r"'\\\%]/g, function (char) {
        switch (char) {
            case "\0":
                return "\\0";
            case "\x08":
                return "\\b";
            case "\x09":
                return "\\t";
            case "\x1a":
                return "\\z";
            case "\n":
                return "\\n";
            case "\r":
                return "\\r";
            case "\"":
            case "'":
            case "\\":
            case "%":
                return "\\" + char; // prepends a backslash to backslash, percent,
                                    // and double/single quotes
        }
    });
}

function dateFormat(data) {
    const array_data = data.split('/');
    if (array_data.length === 3) {
        return array_data[2] + '-' + array_data[1] + '-' + array_data[0];
    }
    return data;
}

function prepareImport(data) {
    const array_sql = [];
    const insert = 'INSERT IGNORE INTO itens VALUES';
    let values = "";
    let row = 1, i = 0;
    const len = data.length;

    let lines = 0;
    for (let a = 0; a < len; a++) {
        row++;
        const dtInicio = dateFormat(data[a][dt_inicio]);
        const dtFim = dateFormat(data[a][dt_fim]);
        const dtGeracao = dateFormat(data[a][dt_geracao]);

        // chave
        let chave = data[a][num_processo] + '#' + data[a][cod_reduzido] + '#' + data[a][seq_item_processo];

        values += "\n(NULL, " + data[a][id_item_processo] + ", " + data[a][id_item_contrato] + ", \"" + data[a][cod_despesa] + "\", \"" + data[a][descr_despesa] + "\", \"" + data[a][descr_tipo_doc] + "\", \"" + data[a][num_contrato] + "\", \"" + data[a][num_processo] + "\", \"" + data[a][descr_mod_compra] + "\", \"" + data[a][num_licitacao] + "\", \"" + dtInicio + "\", \"" + dtFim + "\", \"" + dtGeracao + "\", \"" + data[a][cgc_fornecedor] + "\", \"" + data[a][nome_fornecedor] + "\", \"" + data[a][num_extrato] + "\", \"" + data[a][cod_estruturado] + "\", \"" + data[a][nome_unidade] + "\", \"" + data[a][cod_reduzido] + "\", \"" + data[a][complemento_item] + "\", \"" + data[a][descricao] + "\", \"" + data[a][id_extrato_contr] + "\", \"" + data[a][vl_unitario] + "\", " + data[a][qt_contrato] + ", \"" + data[a][vl_contrato] + "\", " + data[a][qt_utilizado] + ", \"" + data[a][vl_utilizado] + "\", " + data[a][qt_saldo] + ", \"" + data[a][vl_saldo] + "\", \"" + data[a][id_unidade] + "\", \"" + data[a][ano_orcamento] + "\", 0, \"" + chave + "\", \"" + data[a][seq_item_processo] + "\")";

        lines++;

        if (row === 70) {
            array_sql[i] = insert + values + ';';
            values = "";
            i++;
            row = 1;
        } else if (a !== len - 1) {
            values += ", ";
        }
    }
    console.log('importing lines: ' + lines);
    array_sql[i] = insert + values + ';';
    return array_sql;
}

function send(data) {
    CSVsaida.innerHTML += "<p>Enviando. AGUARDE...</p>";
    $.post('php/geral.php', {
        form: 'importItensTest',
        array_sql: data
    }).done(function (resposta) {
        CSVsaida.innerHTML += "Processo finalizado!";
        alert("Resposta do servidor: " + resposta);
    });
}

function leCSV(evt) {
    CSVsaida.innerHTML += "<p>Lendo arquivo...</p>"
    const fileArr = evt.target.result.split('\n');

    CSVsaida.innerHTML += "<p>Linhas encontradas: " + fileArr.length + "</p>";
    let array = [];
    CSVsaida.innerHTML += "<p>Formatando...</p>";
    for (let i = 1; i < fileArr.length; i++) {
        const fileLine = fileArr[i].split('	');
        for (let j = 0; j < fileLine.length; j++) {
            fileLine[j] = fileLine[j].replace("\"", "'");
            fileLine[j] = mysql_real_escape_string(fileLine[j]);
        }
        array[i - 1] = fileLine;
    }

    CSVsaida.innerHTML += "<p>Preparando importação...</p>";
    const array_sql = prepareImport(array);
    CSVsaida.innerHTML += "<p>Preparação concluída</p>";
    send(array_sql);
}